ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c

```diff
@@ -0,0 +1,3367 @@
+#include "opal/class/opal_free_list.h"
+#include "opal/mca/btl/btl.h"
+
+#include "ompi/constants.h"
+#include "ompi/datatype/ompi_datatype.h"
+#include "ompi/datatype/ompi_datatype_internal.h"
+#include "ompi/mca/coll/coll.h"
+#include "ompi/mca/coll/base/coll_tags.h"
+#include "ompi/mca/bml/bml.h"
+#include "ompi/op/op.h"
+#include "ompi/mca/bml/base/base.h"
+
+#include "coll_ftagree.h"
+#include "coll_ftagree_era.h"
+
+
+extern int mca_coll_ftagree_cur_era_topology;
+
+static int era_inited = 0;
+static opal_hash_table_t era_passed_agreements;
+static opal_hash_table_t era_ongoing_agreements;
+static opal_hash_table_t era_incomplete_messages;
+static ompi_comm_rank_failure_callback_t *ompi_stacked_rank_failure_callback_fct = NULL;
+static uint64_t msg_seqnum = 1;
+static opal_free_list_t era_iagree_requests = {{{0}}};
+
+extern int coll_ftagree_debug_rank_may_fail;
+
+typedef enum {
+    MSG_UP = 1,
+    MSG_DOWN,
+    MSG_RESULT_REQUEST
+} era_msg_type_t;
+
+/**
+ * This enum defines the status of processes
+ * when the consensus has not been decided yet
+ */
+enum {
+    /* These values are for the current process */
+    NOT_CONTRIBUTED = 1,
+    GATHERING,
+    BROADCASTING,
+    COMPLETED
+};
+typedef uint32_t era_proc_status_t;
+#define OP_NOT_DEFINED     (1<<31)
+
+typedef struct {
+    union {
+        /** Order is performance critical:
+         *   Hash tables (as of June 11, 2014) consider only the lower x bits for
+         *   hash value (key & mask). As a consequence, having mostly stable entries
+         *   like epoch or contextid at the low part of the 64bits key anihilates the
+         *   performance of the hash tables. The most varying 16 bits should be kept
+         *   first (assuming little endian).
+         */
+        struct {
+            uint16_t agreementid;
+            uint16_t contextid;    /**< Although context ids are 32 bit long, only the lower 16 bits are used */
+            uint32_t epoch;
+        } fields;
+        uint64_t uint64;
+    } u;
+} era_identifier_t;
+
+static inline uint64_t hash_name(opal_process_name_t name) {
+    /** Order is performance critical:
+      *   Hash tables (as of June 11, 2014) consider only the lower x bits for
+      *   hash value (key & mask). As a consequence, having mostly stable entries
+      *   like epoch or contextid at the low part of the 64bits key anihilates the
+      *   performance of the hash tables. The most varying 16 bits should be kept
+      *   first (assuming little endian).
+      */
+    union {
+        struct {
+            uint16_t vpidlo;
+            uint16_t vpidhi;
+            uint32_t jobid;
+        } fields;
+        uint64_t uint64;
+    } hash;
+
+    hash.fields.vpidlo = name.vpid;
+    hash.fields.vpidhi = name.vpid>>16;
+    hash.fields.jobid = name.jobid;
+    return hash.uint64;
+}
+
+#define ERAID_KEY    u.uint64
+#define ERAID_FIELDS u.fields
+
+#if OPAL_ENABLE_DEBUG
+#define PROGRESS_FAILURE_PROB 0.05
```

Comment:
If we want to keep the capability of injecting failures, I'm sure we don't want to force a probability of failure of 0.05 every time somebody defines OPAL_ENABLE_DEBUG.

Reply:
We can leave the error insertion code but it should be protected via a more sensible mechanism (not just DEBUG as Thomas mentionned). Separate this in 2, a define to protect the generator and an MCA param to set the probability.